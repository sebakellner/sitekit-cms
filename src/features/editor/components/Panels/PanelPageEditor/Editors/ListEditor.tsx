import { Box, Button, Field, Input, Separator, VStack } from '@chakra-ui/react'
import type { PanelEditorBase } from '@features/editor/types/panelEditors.types'
import { Plus } from 'lucide-react'

export type ListItem = Record<
  string,
  string | number | boolean | null | undefined
>

const getDefaultItem = (template: Record<string, any>) => {
  const lorem: Record<string, string> = {
    label: 'Lorem Ipsum',
    description: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit.',
  }
  return Object.fromEntries(
    Object.keys(template).map((key) => [key, lorem[key] ?? ''])
  )
}

const ListEditor = ({ value, onChange }: PanelEditorBase<ListItem[]>) => {
  if (!Array.isArray(value)) return null

  const handleChange = (idx: number, key: string, newVal: string) => {
    const updated = value.map((item, i) =>
      i === idx ? { ...item, [key]: newVal } : item
    )
    onChange(updated)
  }

  const handleAdd = () => {
    const newItem = getDefaultItem(value[0])
    onChange([...value, newItem])
  }

  return (
    <VStack align="stretch" gap={1}>
      {value.map((item, idx) => (
        <Box key={idx} mb={2}>
          {Object.entries(item).map(([key, val], idx) => (
            <Field.Root key={idx} mb={3} orientation="horizontal">
              <Field.Label fontSize="xs">
                {key.charAt(0).toUpperCase() + key.slice(1)}
              </Field.Label>
              <Input
                variant="subtle"
                flex="1"
                value={val == null ? '' : String(val)}
                onChange={(e) => handleChange(idx, key, e.target.value)}
              />
            </Field.Root>
          ))}
          <Separator />
        </Box>
      ))}
      <Button variant="outline" onClick={handleAdd}>
        <Plus />
        Add Item
      </Button>
    </VStack>
  )
}

export default ListEditor
