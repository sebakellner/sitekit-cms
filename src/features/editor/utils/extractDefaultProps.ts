import { ComponentMetaSchema } from '@components/site/schemas/componentMeta.schema'
import type { ComponentMeta } from '@components/site/types'

function isDefaultProp(val: unknown): val is { default: unknown } {
  return typeof val === 'object' && val !== null && 'default' in val
}

export function extractDefaultProps(
  props: ComponentMeta['props'] | undefined,
  fallback: unknown = undefined
): Record<string, unknown> {
  if (!props) return {}
  const result = ComponentMetaSchema.shape.props.safeParse(props)
  if (!result.success) {
    console.error('Invalid props schema:', result.error)
    return {}
  }
  const validProps = result.data
  return Object.fromEntries(
    Object.entries(validProps).map(([key, val]) => [
      key,
      isDefaultProp(val) ? val.default : fallback,
    ])
  )
}
