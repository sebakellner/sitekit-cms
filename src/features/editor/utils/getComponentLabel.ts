import React from 'react'

export function getComponentLabel(
  children: React.ReactNode,
  fallback = 'Component'
): string {
  if (React.isValidElement(children)) {
    const type = children.type
    if (typeof type === 'function' && type.name) return type.name
  }
  return fallback
}
