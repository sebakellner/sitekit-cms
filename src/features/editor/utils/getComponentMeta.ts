import { ComponentMetaSchema } from '@components/site/schemas/componentMeta.schema'

type ModuleWithDefault = { default: unknown }

export function getComponentMeta(module: unknown) {
  if (
    module &&
    typeof module === 'object' &&
    'default' in module &&
    (module as ModuleWithDefault).default &&
    typeof (module as ModuleWithDefault).default === 'object'
  ) {
    const def = (module as ModuleWithDefault).default
    const result = ComponentMetaSchema.safeParse(def)
    if (result.success) {
      return result.data
    }
  }
  return { name: '', props: {} }
}
