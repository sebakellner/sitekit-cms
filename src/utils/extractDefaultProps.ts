import type { ComponentMeta } from '@components/site/types'

export function extractDefaultProps(
  props: ComponentMeta['props'] | undefined
): Record<string, unknown> {
  if (!props) return {}
  return Object.fromEntries(
    Object.entries(props).map(([key, val]) => [
      key,
      (val as { default: unknown }).default,
    ])
  )
}
