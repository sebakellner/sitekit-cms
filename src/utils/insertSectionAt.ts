import type { Section } from '@src/types'

export function insertSectionAtUtil(
  sections: Section[],
  setSections: (sections: Section[]) => void,
  id: string,
  index: number
) {
  const sectionId = String(id).startsWith('selector-')
    ? String(id).replace('selector-', '')
    : id
  const uniqueId = `${sectionId}-${Date.now()}`
  const newSection = {
    id: uniqueId,
    name: sectionId,
    props: {},
  }
  const newSections = [...sections]
  newSections.splice(index, 0, newSection)
  setSections(newSections)
}
