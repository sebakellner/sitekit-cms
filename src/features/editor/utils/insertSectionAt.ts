import { componentsRegistry } from '@registry/componentRegistry'
import { v4 as uuidv4 } from 'uuid'

import type { ComponentMeta } from '@components/site/types'
import { extractDefaultProps } from '@src/features/editor/utils/extractDefaultProps'
import type { Section } from '../schemas/section.schema'

export async function insertSectionAtUtil(
  sections: Section[],
  setSections: (sections: Section[]) => void,
  name: string,
  index: number
) {
  const loader = componentsRegistry[name as keyof typeof componentsRegistry]
  if (!loader) {
    console.warn(`Component "${name}" not found in registry.`)
    return
  }

  const metaModule = await loader()

  const meta = metaModule.default as ComponentMeta

  const defaultProps = extractDefaultProps(meta?.props)

  const newSection: Section = {
    id: uuidv4(),
    name,
    props: defaultProps,
  }

  const updatedSections = [
    ...sections.slice(0, index),
    newSection,
    ...sections.slice(index),
  ]

  setSections(updatedSections)
}
