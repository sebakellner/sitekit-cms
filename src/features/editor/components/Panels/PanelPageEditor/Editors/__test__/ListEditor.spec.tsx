import { describe, test, expect } from 'vitest'
import type { PanelEditorBase } from '@features/editor/types/panelEditors.types'
import ListEditor, { type ListItem } from '../ListEditor'
import { render } from 'test/utils/render'
import { screen, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'

import { useState } from 'react'

const setup = (overrides: Partial<PanelEditorBase<ListItem[]>> = {}) => {
  const defaultProps: PanelEditorBase<ListItem[]> = {
    title: 'Color Picker',
    value: [{ id: '1', label: 'Item 1' }],
    onChange: () => {},
  }
  const props = { ...defaultProps, ...overrides }

  const ControlledListEditor = () => {
    const [value, setValue] = useState(props.value)
    const handleChange = (newValue: ListItem[]) => {
      setValue(newValue)
      props.onChange(newValue)
    }
    return <ListEditor {...props} value={value} onChange={handleChange} />
  }

  return <ControlledListEditor />
}

describe('ListEditor', () => {
  test('should match with snapshot', () => {
    const { container } = render(setup())
    expect(container).toMatchSnapshot()
  })

  test('should handle empty list', () => {
    render(setup({ value: [] }))
    expect(screen.queryByLabelText('Label')).not.toBeInTheDocument()
  })

  test('should call onChange with updated list when item is added', async () => {
    const handleChange = vi.fn()
    render(
      setup({
        onChange: handleChange,
      })
    )

    const AddItemBtn = screen.getByRole('button', { name: /Add Item/i })
    expect(AddItemBtn).toBeInTheDocument()

    await userEvent.click(AddItemBtn)

    const labelInputs = screen.getAllByLabelText('Label')
    const newLabelInput = labelInputs[labelInputs.length - 1]

    expect(newLabelInput).toHaveValue('')

    await userEvent.clear(newLabelInput)
    await userEvent.type(newLabelInput, 'Item 2')
    await userEvent.tab()

    await waitFor(() => {
      const calls = handleChange.mock.calls
      expect(calls[calls.length - 1][0]).toEqual([
        { id: '1', label: 'Item 1' },
        { id: '', label: 'Item 2' },
      ])
    })
  })
  test('should call onChange with updated list when item is modified', async () => {
    const handleChange = vi.fn()
    render(
      setup({ value: [{ id: '1', label: 'Item 1' }], onChange: handleChange })
    )
    const labelInput = screen.getByLabelText('Label')
    expect(labelInput).toHaveValue('Item 1')

    await userEvent.clear(labelInput)
    await userEvent.type(labelInput, 'Updated Item 1')
    await userEvent.tab()

    await waitFor(() => {
      const calls = handleChange.mock.calls
      expect(calls[calls.length - 1][0]).toEqual([
        { id: '1', label: 'Updated Item 1' },
      ])
    })
  })
  test('should not crash if value is not an array', () => {
    const { container } = render(setup({ value: undefined }))
    expect(container).toMatchSnapshot()
  })
  test('should handle list items without id field', async () => {
    const handleChange = vi.fn()
    render(
      setup({
        value: [{ label: 'Item 1' }],
        onChange: handleChange,
      })
    )

    const labelInput = screen.getByLabelText('Label')
    expect(labelInput).toHaveValue('Item 1')

    await userEvent.clear(labelInput)
    await userEvent.type(labelInput, 'Updated Item 1')
    await userEvent.tab()

    await waitFor(() => {
      const calls = handleChange.mock.calls
      expect(calls[calls.length - 1][0]).toEqual([{ label: 'Updated Item 1' }])
    })
  })
  test('should render input with empty string if value is null', () => {
    render(setup({ value: [{ id: '1', label: null }] }))
    const labelInput = screen.getByLabelText('Label')
    expect(labelInput).toHaveValue('')
  })

  test('should render input with empty string if value is undefined', () => {
    render(setup({ value: [{ id: '1', label: undefined }] }))
    const labelInput = screen.getByLabelText('Label')
    expect(labelInput).toHaveValue('')
  })

  test('should add item with same fields as first item when list is not empty', async () => {
    render(setup({ value: [{ id: '1', label: 'Item 1', extra: 'foo' }] }))

    const AddItemBtn = screen.getByRole('button', { name: /Add Item/i })
    await userEvent.click(AddItemBtn)

    const labelInputs = screen.getAllByLabelText('Label')
    expect(labelInputs.length).toBe(2)
    expect(labelInputs[1]).toHaveValue('')

    const extraInputs = screen.getAllByLabelText('Extra')
    expect(extraInputs.length).toBe(2)
    expect(extraInputs[1]).toHaveValue('')
  })
  test('should add item with empty fields when list is empty', async () => {
    render(setup({ value: [] }))

    const AddItemBtn = screen.getByRole('button', { name: /Add Item/i })
    await userEvent.click(AddItemBtn)

    expect(screen.queryByLabelText('Label')).not.toBeInTheDocument()
  })
})
