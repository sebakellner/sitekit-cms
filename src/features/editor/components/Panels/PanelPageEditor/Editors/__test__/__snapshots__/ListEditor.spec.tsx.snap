// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ListEditor > should match with snapshot 1`] = `
<div>
  <script
    nonce=""
  >
    ((e,i,s,u,m,a,l,h)=&gt;{let d=document.documentElement,w=["light","dark"];function p(n){(Array.isArray(e)?e:[e]).forEach(y=&gt;{let k=y==="class",S=k&&a?m.map(f=&gt;a[f]||f):m;k?(d.classList.remove(...S),d.classList.add(a&&a[n]?a[n]:n)):d.setAttribute(y,n)}),R(n)}function R(n){h&&w.includes(n)&&(d.style.colorScheme=n)}function c(){return window.matchMedia("(prefers-color-scheme: dark)").matches?"dark":"light"}if(u)p(u);else try{let n=localStorage.getItem(i)||s,y=l&&n==="system"?c():n;p(y)}catch(n){}})("class","theme","system",null,["light","dark"],null,true,true)
  </script>
  <div
    class="chakra-stack css-1qvcor1"
  >
    <div
      class="css-c1sqso"
    >
      <div
        class="chakra-field__root css-1h6z8ek"
        data-part="root"
        data-scope="field"
        id="field::«r0»"
        orientation="horizontal"
        role="group"
      >
        <label
          class="chakra-field__label css-84rnco"
          data-part="label"
          data-scope="field"
          for="«r0»"
          id="field::«r0»::label"
        >
          Id
        </label>
        <input
          class="chakra-input css-1yokl5a"
          data-part="input"
          data-scope="field"
          id="«r0»"
          value="1"
        />
      </div>
      <div
        class="chakra-field__root css-1h6z8ek"
        data-part="root"
        data-scope="field"
        id="field::«r1»"
        orientation="horizontal"
        role="group"
      >
        <label
          class="chakra-field__label css-84rnco"
          data-part="label"
          data-scope="field"
          for="«r1»"
          id="field::«r1»::label"
        >
          Label
        </label>
        <input
          class="chakra-input css-1yokl5a"
          data-part="input"
          data-scope="field"
          id="«r1»"
          value="Item 1"
        />
      </div>
      <span
        aria-orientation="horizontal"
        class="chakra-separator css-1b65dgi"
        role="separator"
      />
    </div>
    <button
      class="chakra-button css-1ocnt5x"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="lucide lucide-plus"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M5 12h14"
        />
        <path
          d="M12 5v14"
        />
      </svg>
      Add Item
    </button>
  </div>
</div>
`;

exports[`ListEditor > should not crash if value is not an array 1`] = `
<div>
  <script
    nonce=""
  >
    ((e,i,s,u,m,a,l,h)=&gt;{let d=document.documentElement,w=["light","dark"];function p(n){(Array.isArray(e)?e:[e]).forEach(y=&gt;{let k=y==="class",S=k&&a?m.map(f=&gt;a[f]||f):m;k?(d.classList.remove(...S),d.classList.add(a&&a[n]?a[n]:n)):d.setAttribute(y,n)}),R(n)}function R(n){h&&w.includes(n)&&(d.style.colorScheme=n)}function c(){return window.matchMedia("(prefers-color-scheme: dark)").matches?"dark":"light"}if(u)p(u);else try{let n=localStorage.getItem(i)||s,y=l&&n==="system"?c():n;p(y)}catch(n){}})("class","theme","system",null,["light","dark"],null,true,true)
  </script>
</div>
`;
